epochStart/shardchain/validatorInfoProcessor.go

```diff
@@ -164,60 +163,74 @@ func (vip *ValidatorInfoProcessor) processAllPeerMiniBlocks(metaBlock *block.Met
 }
 
 func (vip *ValidatorInfoProcessor) computeMissingPeerBlocks(metaBlock *block.MetaBlock) {
-	missingNumber := uint32(0)
+	numMissingPeerMiniblocks := uint32(0)
 	vip.mutMiniBlocksForBlock.Lock()
 
 	for _, mb := range metaBlock.MiniBlockHeaders {
 		if mb.Type != block.PeerBlock {
 			continue
 		}
 
-		vip.allPeerMiniblocks[string(mb.Hash)] = &miniBlockInfo{}
+		vip.mapAllPeerMiniblocks[string(mb.Hash)] = nil
 
 		mbObjectFound, ok := vip.miniBlocksPool.Peek(mb.Hash)
 		if !ok {
-			missingNumber++
+			numMissingPeerMiniblocks++
 			continue
 		}
 
 		mbFound, ok := mbObjectFound.(*block.MiniBlock)
 		if !ok {
-			missingNumber++
+			numMissingPeerMiniblocks++
 			continue
 		}
 
-		vip.allPeerMiniblocks[string(mb.Hash)] = &miniBlockInfo{mb: mbFound}
+		vip.mapAllPeerMiniblocks[string(mb.Hash)] = mbFound
 	}
 
-	vip.numMissing = missingNumber
+	vip.numMissingPeerMiniblocks = numMissingPeerMiniblocks
 	vip.mutMiniBlocksForBlock.Unlock()
 }
 
-func (vip *ValidatorInfoProcessor) retrieveMissingBlocks() error {
+func (vip *ValidatorInfoProcessor) retrieveMissingBlocks() (map[string]*block.MiniBlock, error) {
 	vip.mutMiniBlocksForBlock.Lock()
 	missingMiniblocks := make([][]byte, 0)
-	for mbHash, mbInfo := range vip.allPeerMiniblocks {
-		if mbInfo.mb == nil {
+	for mbHash, mb := range vip.mapAllPeerMiniblocks {
+		if mb == nil {
 			missingMiniblocks = append(missingMiniblocks, []byte(mbHash))
 		}
 	}
-	vip.numMissing = uint32(len(missingMiniblocks))
+	vip.numMissingPeerMiniblocks = uint32(len(missingMiniblocks))
 	vip.mutMiniBlocksForBlock.Unlock()
 
 	if len(missingMiniblocks) == 0 {
-		return nil
+		return nil, nil
 	}
 
 	go vip.requestHandler.RequestMiniBlocks(core.MetachainShardId, missingMiniblocks)
 
 	select {
 	case <-vip.chRcvAllMiniblocks:
-		return nil
-	case <-time.After(time.Second):
-		return process.ErrTimeIsOut
+		return nil, nil
+	case <-time.After(waitTime):
+		return vip.getAllMissingPeerMiniblocks(), process.ErrTimeIsOut
 	}
 }
 
+func (vip *ValidatorInfoProcessor) getAllMissingPeerMiniblocks() map[string]*block.MiniBlock {
+	vip.mutMiniBlocksForBlock.RLock()
+	defer vip.mutMiniBlocksForBlock.RUnlock()
+
+	mapMissingPeerMiniBlocks := make(map[string]*block.MiniBlock)
+	for hash, mb := range vip.mapAllPeerMiniblocks {
+		if mb == nil {
+			mapMissingPeerMiniBlocks[hash] = mb
```

Comment:
A map with <string, nil> will be sent back. Is that ok? Is the map used instead of a []string just to have an easy delete for a hash?

Reply:
Yes, the intention was to send back only a map with hashes of the missing miniblocks(nil) and in trigger the map will help to easy delete with hash

Reply:
Anyway good point here, I refactored to return [][]byte as in trigger another map is created from this slice

Reply:
ok